PEP 590: update

jdemeyer opened the pull request:
Rewrites/clarify some parts of PEP 590 and change some details, adding myself as author.

Main changes:

- Specify that ``kwnames`` must be either ``NULL`` or a non-empty tuple (this is easier for the callee to detect the case of no keyword arguments).
- Allow subclassing the same way as PEP 580.
- Remove ``PyCall_MakeTpCall`` (what's the use case for making this public API?)
- Change ``METH_VECTORCALL`` to mean that the signature is of type ``vectorcall`` (using the actual ``vectorcall`` signature is more natural and faster since you don't need a wrapper at all)
- In ``PyObject_Vectorcall``, allow passing a keyword dict (better to do conversion from dict to tuple in one place, instead of requiring every caller to do that)
- Pass argument vector as `PyObject *const *args` (it's already like that in the implementation)

Details:

- Change the type of ``tp_vectorcall_offset`` to ``Py_ssize_t`` (why ``uintptr_t``? every other offset also has type ``Py_ssize_t``)
- New macro ``PyVectorcall_NARGS(n)`` for ``n & ~PY_VECTORCALL_ARGUMENTS_OFFSET`` (cleaner and more future-proof in case we ever add more special bits)
- New inline function ``PyVectorcall_Function(obj)`` to return the ``vectorcallfunc`` pointer stored in ``obj`` (or `NULL`).
- Change ``PyTupleObject`` in signature to ``PyObject`` (it's already like that in the implementation, passing a specific struct in the Python/C API is normally not done)
- Rename ``PY_VECTORCALL_ARGUMENTS_OFFSET`` -> ``PY_VECTORCALL_PREPEND`` (I didn't like the word "offset" for that, while "prepend" refers to `_PyObject_Call_Prepend`)
- Rename ``vectorcall`` -> ``vectorcallfunc`` (analogous to other function pointer typedefs like `binaryfunc`)
- Rename ``PyCall_MakeVectorCall`` -> ``PyVectorcall_Call``

CC @encukou @markshannon

See also the new implementation at https://github.com/python/cpython/pull/13185

encukou commented:
What's the reason for `PyVectorcall_Check()`? The `_Check` functions are usually used for type checking. Is the type flag not enough?

I'm not convinced about the limitation that `kwnames` can't be the empty tuple. That sounds like an unnecessary limitation.

IMO `Py_TPFLAGS_METHOD_DESCRIPTOR` was a better name: it's concerned with the unbound→bound method mechanism. "C functions" don't this.

IMO `PyCall_MakeVectorCall` was a good name; "CFunction" is a type, but "Vectorcall" is not.

jdemeyer commented:
> it's concerned with the unbound→bound method mechanism.

But Python 3 doesn't really have "unbound methods". It has functions.

jdemeyer commented:
> I'm not convinced about the limitation that `kwnames` can't be the empty tuple. That sounds like an unnecessary limitation.

It's a simplification and optimization. Imagine a C function which doesn't take keywords. With the current PEP 590, it needs to check for that using
```
if (kwnames != NULL && PyTuple_GET_SIZE(kwnames))
```
I suggest to simplify that to
```
if (kwnames != NULL)
```

jdemeyer commented:
> What's the reason for `PyVectorcall_Check()`? The `_Check` functions are usually used for type checking. Is the type flag not enough?

There is no good reason. I just wanted to have a simpler way to write
```
PyType_HasFeature(Py_TYPE(obj), PY_TPFLAGS_HAVE_VECTORCALL)
```

jdemeyer commented:
>  "CFunction" is a type, but "Vectorcall" is not.

Right, but there are other C API functions referring to a protocol like ``PySequence_GetItem`` and ``PyNumber_Add``. I think ``PyVectorcall_Call`` could fit in that.

encukou commented on pep-0590.rst:
```diff
@@ -48,145 +48,212 @@ The function pointer type
 
-``Py_TPFLAGS_METHOD_DESCRIPTOR`` should be set if the the callable uses the descriptor protocol to create a method or method-like object.
-This is used by the interpreter to avoid creating temporary objects when calling methods.
+- ``func.__set__`` and ``func.__delete__`` must not be implemented.
```
Why must they not be implemented?
The object must ensure the `Py_TPFLAGS_FUNCTION_DESCRIPTOR` invariants hold even after `__set__` or `__delete__` is run, which might be hard, but I don't see a reason to disallow them.

jdemeyer replied:
Agreed. I copied that from PEP 580 but there is no fundamental reason to disallow `__set__`/`__delete__`.

encukou commented on pep-0590.rst:
```diff
@@ -48,145 +48,212 @@ The function pointer type
+See [3]_ for an example of how ``PY_VECTORCALL_PREPEND`` is used by a callee to avoid allocation.
+
+For getting the actual number of arguments from the parameter ``n``,
+the macro ``PyVectorcall_NARGS(n)`` should be used.
```
```suggestion
the macro ``PyVectorcall_NARGS(n)`` must be used.
```

encukou commented on pep-0590.rst:
```diff
@@ -48,145 +48,212 @@ The function pointer type
+  This is mostly meant to put in the ``tp_call`` slot.
 
-Both functions raise an exception if ``obj`` does not support the relevant protocol.
+- ``Py_ssize_t PyVectorcall_NARGS(Py_ssize nargs)``: currently equivalent to
```
```suggestion
- ``Py_ssize_t PyVectorcall_NARGS(Py_ssize nargs)``: Given a vectorcall ``nargs`` argument, return the actual number of arguments. Currently equivalent to
```

encukou commented on pep-0590.rst:
```diff
@@ -213,30 +280,25 @@ Removing any special cases and making all calls use the ``tp_call`` form was als
 ================
 
-Victor Stinner for developing the original "vector call" calling convention internally to CPython (where is it is called "fast call")
+Victor Stinner for developing the original "fastcall" calling convention internally to CPython (where is it is called "fast call")
```
```suggestion
Victor Stinner for developing the original "vector call" calling convention internally to CPython (where is it is called "fast call")
```
Either leave this be, or remove the note in parentheses.

markshannon replied:
Victor calls it "fastcall" not "vector call".

encukou commented:
### NULL vectrorcall

> No longer allow the vectorcall pointer in the instance to be NULL (for simplicity)

Unfortunately, that makes the type optimization impossible -- we want `list` to have vectorcall, but not all the other types.

### METHOD_DESCRIPTOR/FUNCTION_DESCRIPTOR

But Python 3 does have unbound methods:

```
>>> list.append
<method 'append' of 'list' objects>
```

"method" describes the mechanism better than "function".

###  limitation that kwnames can't be the empty tuple

You make a fair point. But let's enforce this, at least in debug mode.

### PyVectorcall_Check

Since I'm currently at PyCon sprints, I asked Victor for another point of view here. He agrees that it's not necessary.
Please don't add the extra function. Use `PyType_HasFeature` instead.

### PyVectorcall_Call / PyCall_MakeVectorCall

Let's keep `PyCall` as the namespace here. Maybe `PyCall_VectorCall` would work, too.

jdemeyer commented on pep-0590.rst:
```diff
@@ -11,34 +11,34 @@ Post-History:
-
-The current ``tp_call`` calling convention requires creation of a tuple and, if there are any named arguments, a dictionary for every call.
-This is expensive. The proposed calling convention removes the need to create almost all of these temporary objects.
-Another source of inefficiency in the ``tp_call`` convention is that it has one function pointer per-class, rather than per-object. This is inefficient for calls to classes as several intermediate objects need to be created. For a user defined class, ``UserClass``, at least one intermediate object is created for each call in the sequence ``type.__call__``, ``object.__new__``, ``UserClass.__init__ ...
```
It's not really removed. I just grouped everything regarding `type.__call__` in one section.

jdemeyer commented:
> we want `list` to have vectorcall, but not all the other types.

Is there a problem with using a generic vectorcall wrapper for all types? This could be set up by `PyType_Ready`. It's just an idea, feel free to say that it sucks :-)

markshannon commented on pep-0590.rst:
```diff
@@ -48,145 +48,212 @@ The function pointer type
-The new flag specifies that the function has the type ``PyObject *(*call) (PyObject *self, PyObject *const *args, Py_ssize_t nargs, PyObject *kwname)``
+A new constant ``METH_VECTORCALL`` is added for specifying ``PyMethodDef`` structs.
+It means that the C function has the signature ``vectorcallfunc``.
+This should be the preferred flag for new functions, as this avoids a wrapper function.
```
This is a major change and prevents the interpreter from sanity checking the return value of `PyMethodDef` defined functions. The signature of `METH_VECTORCALL` should remain as
`PyObject *(*call) (PyObject *self, PyObject *const *args, Py_ssize_t nargs, PyObject *kwname)`

jdemeyer replied:
> This is a major change and prevents the interpreter from sanity checking the return value of `PyMethodDef` defined functions.

Please explain. I don't think it's the job of the vectorcall **callee** to do that check, it should be done by the vectorcall **caller**.

markshannon replied:
No, it is the job of the callee to do this check. `_PyCFunction_FastCallKeywords` does this, as the `PymethodDef` defined code is not entirely trusted. However, calls to the interpreter or builtin classes *are* trusted and don't need to be double checked.

jdemeyer replied:
> as the `PymethodDef` defined code is not entirely trusted.

Why should the `PyMethodDef`-defined code be less trusted than other code called by the vectorcall protocol? I don't see the difference.

markshannon replied:
Because `PyMethodDef` remains the preferred way for builtin functions to be implemented if they are implemented in C. For tools like Cython that can ensure that the returned value is consistent, the additional check is not required.

markshannon commented on pep-0590.rst:
```diff
@@ -48,145 +48,212 @@ The function pointer type
-------------------------------------------------------------------
+For ``builtin_function_or_method`` and ``method_descriptor``
+(which use the ``PyMethodDef`` data structure),
+a specific vectorcall wrapper will be implemented for every existing calling convention.
```
We could do this, but it might not be worth it. We shouldn't specify it in the PEP.

jdemeyer replied:
I think it's an interesting idea, so I would keep it in the PEP. But I'll just say that it's something which "could" be done, not "will" be done.

markshannon commented on pep-0590.rst:
```diff
@@ -48,145 +48,212 @@ The function pointer type
+requires multiple calls.
+At least one intermediate object is created for each call in the sequence
+``type.__call__``, ``cls.__new__``, ``cls.__init__``.
+So it makes a lot of sense to use vectorcall also for calling classes.
```
Remove "also"

markshannon commented on pep-0590.rst:
```diff
@@ -48,145 +48,212 @@ The function pointer type
 Argument Clinic [4]_ automatically generates wrapper functions around lower-level callables, providing safe unboxing of primitive types and
 other safety checks.
-Argument Clinic could be extended to generate wrapper objects conforming to the new ``vectorcall`` protocol.
+Argument Clinic could be extended to generate wrapper objects with the ``METH_VECTORCALL`` signature.
```
Leave this as it is. `METH_VECTORCALL` is not equivalent to `vectorcall`.

markshannon commented on pep-0590.rst:
```diff
@@ -48,145 +48,212 @@ The function pointer type
+
+The motivation is also different: this PEP is mainly meant to
+allow optimizating external classes (outside of CPython),
+not to optimize CPython itself.
```
This PEP is meant to allow optimization of CPython as well.

jdemeyer replied:
You mean specifically for `type` or do mean something else also?

encukou commented on pep-0590.rst:
```diff
@@ -48,145 +48,212 @@ The function pointer type
+--------------------------------------
 
-The unused slot ``printfunc tp_print`` is replaced with ``tp_vectorcall_offset``. It has the type ``uintptr_t``.
+The unused slot ``printfunc tp_print`` is replaced with ``tp_vectorcall_offset``. It has the type ``Py_ssize_t``.
```
Specifically, we can use [some C99 features](https://www.python.org/dev/peps/pep-0007/#c-dialect). But `stdint.h`'s `uintptr_t` is one of them.
Older code doesn't use it because the C99 features were allowed relatively recently.

jdemeyer replied:
In which sense is `uintptr_t` more "correct" than `Py_ssize_t`? Given that all other offset values (for example `tp_weaklistoffset`) use `Py_ssize_t`, there should be a good reason to deviate from that.

larryhastings replied:
To be pedantic, neither `uintptr_t` nor `Py_ssize_t` is correct.  The C standard says you're not allowed to store the address of a function in a variable not declared as a pointer to a function.  We do it all the time, of course.  But it would do my heart some good to not add *new* places where we bend the spec.

markshannon replied:
It is not an address, but an offset. 
However, as we want to minimize the chance of breakage and the offset will always be less than `UINTPTR_MAX`, `uintptr_t` is better than `offset_t`.

markshannon replied:
`Py_ssize_t` exists only because our C code had to be C89 compliant.

jdemeyer replied:
To be even more pedantic, wouldn't `ptrdiff_t` be better? It's specifically meant for differences between pointers (i.e. offsets).

jdemeyer replied:
FYI: according to https://en.cppreference.com/w/c/types/offsetof the resulting type of `offsetof()` is a `size_t`, so using `size_t` for `tp_vectorcall_offset` can't be wrong. And CPython typically uses `Py_ssize_t` instead of `size_t` for sizes, so `Py_ssize_t` is not that wrong either.

I don't want further bikeshedding on this, so please suggest something and I'll make the change.

jdemeyer commented on pep-0590.rst:
```diff
@@ -48,145 +48,212 @@ The function pointer type
+* ``Py_ssize_t n``: The number of arguments plus the optional flag ``PY_VECTORCALL_PREPEND`` (see below)
 * ``PyObject **args``: A vector of arguments
-* ``PyTupleObject *kwnames``: A tuple of the names of the named arguments.
+* ``PyObject *kwnames``: Either ``NULL`` or a non-empty tuple of the names of the keyword arguments
```
Why? The protocol is easier to use if there is a unique way to specify "no keyword arguments".

scoder replied:
I don't have a strong opinion here, but I agree with Jeroen. The caller probably knows already if the tuple is empty or not, so why not help the callee here? As long as `PyObject_VectorCall()` is used for calling into vectorcallees, it can normalise that case internally, and everyone who does not use it is on their own anyway.

jdemeyer commented on pep-0590.rst:
```diff
@@ -48,145 +48,212 @@ The function pointer type
-If this flag is set for a class ``F``, then instances of that class are expected to behave the same as a Python function when used as a class attribute.
-Specifically, this means that the value of ``c.m`` where ``C.m`` is an instance of the class ``F`` (and ``c`` is an instance of ``C``)
-must be an object that acts like a bound method binding ``C.m`` and ``c``.
-This flag is necessary if custom callables are to be able to behave like Python functions *and* be called as efficiently as Python or built-in functions.
```
None, I replaced it with a different description that I find easier to understand. I found your wording hard to read.

encukou replied:
I agree with Jeroen here; the description in terms of `__get__` is clearer. Mark, let me know if you want to keep the discussion open.

jdemeyer commented on pep-0590.rst:
```diff
@@ -11,34 +11,34 @@ Post-History:
 This PEP proposes that the calling convention used internally for Python and builtin functions is generalized and published
 so that all calls can benefit from better performance.
-
-Improved Performance
```
The first two sentences of that section were just repeating stuff which was already said. But I'll keep the part about `type.__call__`.

jdemeyer commented on pep-0590.rst:
```diff
@@ -11,34 +11,34 @@ Post-History:
 
-This PEP introduces a new calling convention [1]_ for use by CPython and other software and tools in the CPython ecosystem.
-The new calling convention is a formalisation and extension of "fastcall", a calling convention already used internally by CPython.
+This PEP introduces a new protocol for fast calling of objects.
```
Maybe it's a calling convention as well as a protocol to use that new calling convention.

markshannon commented:
> > I'm not convinced about the limitation that `kwnames` can't be the empty tuple. That sounds like an unnecessary limitation.
> 
> It's a simplification and optimization. Imagine a C function which doesn't take keywords. With the current PEP 590, it needs to check for that using
> 
> ```
> if (kwnames != NULL && PyTuple_GET_SIZE(kwnames))
> ```
> 
> I suggest to simplify that to
> 
> ```
> if (kwnames != NULL)
> ```

Although simpler, it is no faster the additional check is only required during error handling.
Overall, allowing empty tuples is more general and no slower.

jdemeyer commented on pep-0590.rst:
```diff
@@ -48,145 +48,212 @@ The function pointer type
 
 This is implemented by the function pointer type:
-``typedef PyObject *(*vectorcall)(PyObject *callable, Py_ssize_t n, PyObject** args, PyTupleObject *kwnames);``
+``typedef PyObject *(*vectorcallfunc)(PyObject *callable, Py_ssize_t n, PyObject **args, PyObject *kwnames);``
```
> Although simpler, it is no faster the additional check is only required during error handling.

Not only during error handling. It's relevant for functions that can take keyword arguments but which are called without keyword arguments.

> Overall, allowing empty tuples is more general and no slower.

What would "more general" even mean here? The protocol should be as simple to use as possible and not allowing empty tuples makes it simpler.

What's wrong with disallowing an empty tuple?

jdemeyer replied:
It's just clearer to use a name ending in `func`. CPython has many such names for function pointer types ending in `func`, for example
```
typedef PyObject *(*wrapperfunc)(PyObject *self, PyObject *args,
                                 void *wrapped);
```

encukou replied:
Disallowing an empty tuple adds a non-obvious edge case.
What's wrong with non-obvious edge cases? They rely on either everyone implementing the protocol reading the docs and noticing this case, or on checking for the case and enforcing it. In the ideal world, that would be enough. But if we have a choice, and it doesn't hurt performance, let's not add one.

jdemeyer replied:
> Disallowing an empty tuple adds a non-obvious edge case.

I argue the oppsite: **allowing** an empty tuple adds a non-obvious edge case.

We should really view this from the side of the callee (the C function), as that's what external projects will implement. From that viewpoint, having two ways to indicate "no keywords" is a non-obvious edge case.

The side of the caller (which should ensure not to send an empty tuple) is CPython and there the issue of people implementing the protocol wrongly doesn't arise. External C code is not expected to manually use `tp_vectorcall_offset` to make vectorcalls: it is expected to use an API like `PyCall_Vectorcall()` and that API will ensure to replace an empty tuple with `NULL`.

I see it as an application of https://en.wikipedia.org/wiki/Robustness_principle (Be conservative in what you send, be liberal in what you accept): `PyCall_Vectorcall` should accept an empty tuple but it should not send an empty tuple to the vectorcall function.

jdemeyer commented:
> Although simpler, it is no faster the additional check is only required during error handling.

Not only during error handling. It's relevant for functions that can take keyword arguments but which are called without keyword arguments.

> Overall, allowing empty tuples is more general and no slower.

What would "more general" even mean here? The protocol should be as simple to use as possible and not allowing empty tuples makes it simpler.

What's wrong with disallowing an empty tuple?

jdemeyer commented:
I pushed an update with some of the suggested edits.

encukou commented:
Workflow/process note: in the future, please don't force-push to PRs. It makes the changes harder to review.
The PR will be squashed into one commit when it's merged.

jdemeyer commented:
Mark, Petr: do you agree with all the changes that you did not comment on?

jdemeyer commented:
> Is there a problem with using a generic vectorcall wrapper for all types? This could be set up by `PyType_Ready`. It's just an idea, feel free to say that it sucks :-)

I changed my mind here. Let's allow `NULL` for simplicity.

I would also like to add a macro `PyVectorcall_FUNC(obj)` to return the `vectorcallfunc` pointer stored inside `obj` (which may be `NULL` if `obj` doesn't support vectorcall or if the pointer happens to be `NULL`).

I already made these changes in a new commit.

jdemeyer commented:
I replaced `PyObject **args` by `PyObject *const *args`. I believe this is what we decided already and it's like that anyway in the implementation.

encukou commented:
Putting the non-controversial changes in one PR first, and getting them merged, should help.
Jeroen, I'd be happy to do it, just let me know so we don't duplicate effort.

jdemeyer commented:
> ### PyVectorcall_Call / PyCall_MakeVectorCall
> 
> Let's keep `PyCall` as the namespace here. Maybe `PyCall_VectorCall` would work, too.

I'd like to come back to this. While working on the implementation, I found the name `PyCall_Vectorcall` confusing. C API functions are typically named as
```
{type or protocol}_{operation}
```
For example:
- `PyNumber_Add` uses the number protocol to do the operation `Add`
- `PyCFunction_Call` acts on `PyCFunction` instances and does the operation `Call` (which means taking an `args` tuple and `kwargs` dict).

Now the function under discussion executes the operation `Call` (it takes an `args` tuple and `kwargs` dict) and it does that using the vectorcall protocol. So `PyVectorcall_Call` really seems to be the best name.

jdemeyer commented:
At this point, it's not clear to me what are the controversial changes (for example, I addressed some comments by you two yesterday but I don't know if you're happy with those changes).

As far as I can tell, the major points under discussion are:
- allowing an empty tuple
- meaning of `METH_VECTORCALL`

Then some bikeshedding:
- naming `vectorcall` or `vectorcallfunc`
- naming `PyCall_MakeVectorCall` or `PyCall_VectorCall` or `PyVectorcall_Call`
- type of `tp_vectorcall_offset` (`Py_ssize_t` or `uintptr_t` or ...)

Is there something not on this list that you disagree with?

encukou commented:
I tried to compile a list of the changes here (and which ones are settled).
But the PR keeps changing, so the list is probably outdated now.
This PR is not easy to review :(

* [x] Renaming the PEP
* [x] Listing Jeroen as the author
* [x] Rewording the Abstract
* [x] Abstract note that this doesn't affect the language
* [x] Rename "Rationale" to "Motivation"
* [x] Moving & rewording Motivation section about types
* [x] Renaming "arguments offset" to "prepend"
* [ ] Kwarg name tuple must be non-empty
* [*] `PyObject **args` ⇒ `PyObject *const *args`
* [*] `PyTupleObject *kwnames` → `PyObject *kwnames`
* [ ] `vectorcall` → `vectorcallfunc`
* [ ] offset: `uintptr_t` → `Py_ssize_t`
* [*] Rewording section on `Py_TPFLAGS_HAVE_VECTORCALL`
* [ ] Replacing the `Py_TPFLAGS_METHOD_DESCRIPTOR` specification
* [*] `PyVectorcall_NARGS` macro
* [*] Removing restriction on base classes & documenting what happens
* [*] Rewording the "New C API and changes to CPython" list
* [ ] `METH_VECTORCALL` signature change
* [*] `METH_VECTORCALL` is preferred as it avoids a wrapper function
* [*] Note that `METH_VECTORCALL` is unspecified
* [ ] Adding `wrapper_descriptor` and `method-wrapper` to the list of things that do vectorcall
* [ ] Convention-specific wrappers for `builtin_function_or_method` and `method_descriptor`
* [*] "Using the vectorcall protocol for classes" section
* [ ] Argument Clinic should use `METH_VECTORCALL`, not `vectorcall`
* [*] "Third-party extension classes using vectorcall" section
* [ ] Rewording "Performance implications of these changes"
* [*] Empty "Stable ABI" section
* [ ] bpo-29259 in Alternative Suggestions
* [*] Rewording Victors' "fastcall" acknowledgement

jdemeyer commented:
> Putting the non-controversial changes in one PR first, and getting them merged, should help.
> Jeroen, I'd be happy to do it, just let me know so we don't duplicate effort.

I don't plan any further changes to this PR for now. If it's clear to you what's controversial and what not, feel free to make those changes.

encukou commented:
> If it's clear to you what's controversial and what not, feel free to make those changes.

Alright. It'll take me some time to sort through this.

jdemeyer commented:
Thanks for that list. I'll briefly comment on the unchecked boxes:

> * [ ]  Kwarg name tuple must be non-empty

Being discussed already, you know my opinion. Good to revert for now if you don't agree.

> * [ ]  `vectorcall` → `vectorcallfunc`

Bikeshedding, I don't care much.

> * [ ]  offset: `uintptr_t` → `Py_ssize_t`

Bikeshedding, I don't care much.

> * [ ]  Replacing the `Py_TPFLAGS_METHOD_DESCRIPTOR` specification

Note that the actual specification is not changed, only the wording. I changed it because I found the original text unclear. Feel free to do whatever you want with this.

> * [ ]  `METH_VECTORCALL` signature change

Under discussion, would be good to revert for now.

> * [ ]  Adding `wrapper_descriptor` and `method-wrapper` to the list of things that do vectorcall
>
> * [ ]  Convention-specific wrappers for `builtin_function_or_method` and `method_descriptor`

Both of these are meant as optional things that could possibly be done as part of the implementation but not strictly part of the PEP. They won't be in the initial implementation and they may or may not be done later. I'd still like to keep them in the PEP if only to have some place where these ideas are written down. You could group these in a section "Possible future CPython changes" or something.

> * [ ]  Argument Clinic should use `METH_VECTORCALL`, not `vectorcall`

Related to the discussion about `METH_VECTORCALL`, can be reverted together with the change to `METH_VECTORCALL`.

> * [ ]  Rewording "Performance implications of these changes"

I didn't want to claim performance improvements without evidence. Maybe the best would be to remove this section completely?

> * [ ]  bpo-29259 in Alternative Suggestions

What's controversial about this? bpo-29259 is very close to PEP 590, so it would be good to mention it somewhere.

encukou commented:
> I'd still like to keep them in the PEP if only to have some place where these ideas are written down. 

I agree with using the PEP for listing the "optional things we'd like to get into Python 3.8 if we can make it" (as long as they're marked as optional).

encukou commented:
All the points are covered in GH-1035 now; I'll keep track of them until they're resolved. Discussions can continue here.

jdemeyer commented:
About `METH_VECTORCALL`
=======================

The reason for me to change the meaning of `METH_VECTORCALL` was motivated by the part about [Argument Clinic](https://www.python.org/dev/peps/pep-0590/#the-pymethoddef-protocol-and-argument-clinic) (AC).

AC works by generating code and a `PyMethodDef` entry. We all agree that it's a good idea to have AC-generated code use the vectorcall protocol directly without additional wrappers (this is in the PEP). But if you want to do that, you need a way to indicate that in `PyMethodDef`. So in any case, you need something like my proposal for the `METH_VECTORCALL` flag.

So we need that functionality and I see no reason why it should be reserved for AC, we might as well make it public (and then we can of course bikeshed about the name...).

encukou commented:
>  I see no reason why [METH_VECTORCALL with vectorcallfunc signature] be reserved for AC

To make the `callable` argument usable, all callees would need to get `self` directly out of `PyCFunctionObject`, which is currently an undocumented implementation detail.
Making `PyCFunctionObject` part of the API means we can't change it in the future.

jdemeyer commented:
> which is currently an undocumented implementation detail.

It's undocumented but it's actually officially part of the stable ABI (remember the discussion in PEP 580 about deprecating `PyCFunction_GetFlags`). So it's not an implementation detail.

encukou commented:
`PyCFunction_GetSelf` is part of the stable ABI (for introspection, like `[].append.__self__`). But the C function doesn't have access to the callable object, so it won't use `PyCFunction_GetSelf`.
Making this the preferred way to write extension functions would be a major change.

jdemeyer commented:
> But the C function doesn't have access to the callable object, so it won't use `PyCFunction_GetSelf`.

Yes, it does have access to the callable object, since it's called with the `vectorcallfunc` signature. So the `METH_VECTORCALL` C function would look like

```C
static PyObject *
myfunc(PyObject *callable, PyObject *const *args,
       Py_ssize_t nargs, PyObject *kwnames)
{
    PyObject *self = PyCFunction_GetSelf(callable);
    nargs = PyVectorcall_NARGS(nargs);

    /* From this point on, everything is exactly the same as
     * METH_FASTCALL|METH_KEYWORDS */
}
```

> Making this the preferred way to write extension functions would be a major change.

I'm not saying that this should be the preferred way. I'm just saying: why not allow it? We need the functionality anyway for Argument Clinic, so it's not additional effort to support it.

jdemeyer commented:
Actually, scratch that. You need to support two types (`builtin_function_or_method` and `method_descriptor`), so it's more complicated than that.

So now I would suggest the minimal approach: leave `METH_FASTCALL` exactly as it is currently and don't introduce a new `METH_VECTORCALL` in the PEP (but we do need it privately)

jdemeyer commented:
> Could this PR be split into smaller PRs, please? It makes it a lot easier to merge if distinct changes are in different PRs.

Done now. Closing this PR as everything is now on separate PRs.
